gnoi/reset/server_test.go

```diff
@@ -0,0 +1,54 @@
+package reset
+
+import (
+	"context"
+	"testing"
+
+	"github.com/google/gnxi/gnoi/reset/pb"
+)
+
+func TestStart(t *testing.T) {
+	testSettings := []Settings{
+		{false, false},
+		{false, true},
+		{true, false},
+		{true, true},
+	}
+
+	testRequests := []pb.StartRequest{
+		{ZeroFill: true},
+		{ZeroFill: true, FactoryOs: true},
+		{FactoryOs: true},
+		{},
+	}
+
+	for _, test := range testSettings {
+		s := NewServer(&test)
+		for _, testReq := range testRequests {
+			resp, _ := s.Start(context.Background(), &testReq)
+			switch response := resp.Response.(type) {
+			case *pb.StartResponse_ResetSuccess:
+				if test.errorIfZero && testReq.ZeroFill || test.osUnsupported && testReq.FactoryOs {
+					t.Errorf(
+						"Error occured on case: \nSettings{errIfZero:%v,osUnsupported:%v}\nRequest{ZeroFill:%v,FactoryOs:%v} \nResponseSuccess{%v},",
+						test.errorIfZero,
+						test.osUnsupported,
+						testReq.ZeroFill,
+						testReq.FactoryOs,
+						response)
+				}
+			case *pb.StartResponse_ResetError:
+				if response.ResetError.ZeroFillUnsupported != (test.errorIfZero && testReq.ZeroFill) ||
+					response.ResetError.FactoryOsUnsupported != (test.osUnsupported && testReq.FactoryOs) {
+					t.Errorf(
+						"Error occured on case: \nSettings{errIfZero:%v,osUnsupported:%v}\nRequest{ZeroFill:%v,FactoryOs:%v} \nResponseError{%v},",
+						test.errorIfZero,
+						test.osUnsupported,
+						testReq.ZeroFill,
+						testReq.FactoryOs,
+						response)
+				}
+			}
+		}
+	}
+}
```

Comment:
I think this works, however the double iteration makes it hard to read. I was thinking you could use t.Run and do something like the following (context: https://gobyexample.com/testing):


```suggestion
var tests []struct{
	name string
	request *pb.StartRequest
	settings Settings
}

func makeTests() tests{
	reqs := []pb.StartRequest{
		{ZeroFill: true},
		{ZeroFill: true, FactoryOs: true},
		{FactoryOs: true},
		{},
	}
	sets := []Settings{
		{false, false},
		{false, true},
		{true, false},
		{true, true},
	}
	t := tests
	for _, set := range sets {
		for _, req := range reqs {
		        name = <compute a name based on the current values>
			t = append(t, {name: "some name", request: req, settings: set})
		}
	}
	return t
}

tests := makeTests()
for _, test := range tests {
	s := NewServer(&test)
	t.Run(test.name, func(t *testing.T) {
		resp, err := s.Start(context.Background(), &testReq)
		if err == nil {
			t.Fatalf("ops")
		}
		switch response := resp.Response.(type) {
		case *pb.StartResponse_ResetSuccess:
			if test.errorIfZero && testReq.ZeroFill || test.osUnsupported && testReq.FactoryOs {
				t.Errorf(
					"Error occured on case: \nSettings{errIfZero:%v,osUnsupported:%v}\nRequest{ZeroFill:%v,FactoryOs:%v} \nResponseSuccess{%v},",
					test.errorIfZero,
					test.osUnsupported,
					testReq.ZeroFill,
					testReq.FactoryOs,
					response)
			}
		case *pb.StartResponse_ResetError:
			if response.ResetError.ZeroFillUnsupported != (test.errorIfZero && testReq.ZeroFill) ||
				response.ResetError.FactoryOsUnsupported != (test.osUnsupported && testReq.FactoryOs) {
				t.Errorf(
					"Error occured on case: \nSettings{errIfZero:%v,osUnsupported:%v}\nRequest{ZeroFill:%v,FactoryOs:%v} \nResponseError{%v},",
					test.errorIfZero,
					test.osUnsupported,
					testReq.ZeroFill,
					testReq.FactoryOs,
					response)
			}
		}
	})
}
```

Disclaimer, I haven't tested this :)

optional:
On top of my suggestion I would encourage you to have a want field inside your test struct and compute its value inside the make Tests func. Rationale: We should try to keep the actual test  as clean as possible.